tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc

```diff
@@ -1246,6 +1218,121 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  // * Length of explicit_paddings attribute is valid and has non negative
+  //   elements
+  // * strides and dilations attributes have positive elements
+  if (!IsOfRankOrUnranked(input, num_dims) ||
+      !IsOfRankOrUnranked(filter, num_dims))
+    return emitOptionalError(location, "requires operands to be ", num_dims,
+                             "D tensor");
+
+  if (padding == tensorflow::Padding::EXPLICIT) {
+    if (explicit_padding.size() == 0) {
+      return emitOptionalError(location,
+                               "requires attribute 'explicit_paddings' with "
+                               "'EXPLICIT' padding mode");
+    }
+    if (explicit_padding.size() != num_dims * 2) {
+      return emitOptionalError(
+          location, "requires explicit_paddings attribute length to be ",
+          num_dims * 2);
+    }
+    auto is_negative = [](Attribute val) {
+      return val.cast<IntegerAttr>().getValue().getSExtValue() < 0;
+    };
+    if (llvm::any_of(explicit_padding, is_negative))
+      return emitOptionalError(location,
+                               "requires non negative explicit paddings");
+  }
+
+  if (failed(
+          VerifyConvOpAttributesV2(num_dims, strides, dilations, location))) {
+    return failure();
+  }
+
+  // For operands having dynamic shape.
+  SmallVector<int64_t, 4> return_shape(num_dims, ShapedType::kDynamicSize);
+  if (!input_ty.hasStaticShape() || !filter_ty.hasStaticShape()) {
+    inferredReturnTypes.assign(
+        {RankedTensorType::get(return_shape, input_ty.getElementType())});
+    return success();
+  }
+
+  // Checks the size of each of the output dimension.
+  for (auto i : llvm::seq<int>(0, num_spatial_dims)) {
+    const int64_t dim = GetTensorSpatialDimIndex(num_dims, format, i);
+    int64_t stride = get_int(strides[dim]);
+    tensorflow::int64 expected_output_size;
+    tensorflow::int64 pad_low;
+    tensorflow::int64 pad_high;
+    // Retrieve padding, if defined explicitly.
+    if (padding == tensorflow::Padding::EXPLICIT) {
+      pad_low = get_int(explicit_padding[2 * dim]);
+      pad_high = get_int(explicit_padding[2 * dim + 1]);
+    }
+    // Calculate the expected_output_size.
+    tensorflow::Status status = tensorflow::GetWindowedOutputSizeVerboseV2(
+        input_ty.getDimSize(dim), filter_ty.getDimSize(i),
+        get_int(dilations[dim]), stride, padding, &expected_output_size,
+        &pad_low, &pad_high);
+    // Return failure if expected_output_size could not be calculated.
+    if (!status.ok()) return failure();
+    return_shape[dim] = expected_output_size;
+  }
+
+  // The remaining dimensions can be obtained using utilities from
+  // tensorflow/core/util/tensor_format.h.
+  return_shape[GetTensorBatchDimIndex(num_dims, format)] =
+      input_ty.getShape()[GetTensorBatchDimIndex(num_dims, format)];
+  return_shape[GetTensorFeatureDimIndex(num_dims, format)] =
+      std::is_same<OpT, Conv2DOpAdaptor>() ? filter_ty.getShape()[3]
```

Comment:
optional nit: Instead of using magic constants, we should use GetFilterTensorInnerInputChannelsDimIndex here.